ee/backend/app/src/main/kotlin/io/tolgee/ee/service/LlmProviderService.kt

```diff
@@ -0,0 +1,297 @@
+import kotlin.jvm.optionals.getOrNull
+import kotlin.math.roundToInt
+
+@Service
+class LlmProviderService(
+  private val llmProviderRepository: LlmProviderRepository,
+  private val organizationService: OrganizationService,
+  private val llmPropertiesService: LlmPropertiesService,
+  private val openaiApiService: OpenaiApiService,
+  private val tolgeeApiService: TolgeeApiService,
+  private val ollamaApiService: OllamaApiService,
+  private val cacheManager: CacheManager,
+  private val currentDateProvider: CurrentDateProvider,
+  private val claudeApiService: ClaudeApiService,
+  private val geminiApiService: GeminiApiService,
+  private val restTemplateBuilder: RestTemplateBuilder,
+  private val internalProperties: InternalProperties,
+) {
+  private val cache: Cache by lazy { cacheManager.getCache(Caches.LLM_PROVIDERS) }
+  private var lastUsedMap: MutableMap<String, Long> = mutableMapOf()
+
+  fun getProviderByName(
+    organizationId: Long,
+    name: String,
+    priority: LlmProviderPriority?,
+  ): LlmProviderDto {
+    val customProviders = getAll(organizationId)
+    val serverProviders = getAllServerProviders()
+    val providersOfTheName =
+      if (customProviders.find { it.name == name } != null) {
+        customProviders
+      } else {
+        serverProviders
+      }.filter {
+        it.name == name
+      }
+
+    val providersWithPriority =
+      if (providersOfTheName.find { it.priority == priority } != null) {
+        providersOfTheName.filter { it.priority == priority }
+      } else {
+        providersOfTheName
+      }
+
+    if (providersWithPriority.isEmpty()) {
+      throw BadRequestException(Message.LLM_PROVIDER_NOT_FOUND, listOf(name))
+    }
+
+    val providerInfo = cache.get(name, ProviderInfo::class.java)
+    val providers =
+      providersWithPriority.filter {
+        if (providerInfo != null) {
+          providerInfo.suspendMap.getOrDefault(it.id, 0L) < currentDateProvider.date.time
+        } else {
+          true
+        }
+      }
+
+    if (providers.isEmpty() && providerInfo?.suspendMap?.isNotEmpty() != null) {
+      val closestUnsuspend = providerInfo.suspendMap.map { (_, time) -> time }.min()
+      throw TranslationApiRateLimitException(closestUnsuspend)
+    }
+
+    var lastUsed = lastUsedMap.get(name)
+    val lastUsedIndex = providers.indexOfFirst { it.id == lastUsed }
+    val newIndex = (lastUsedIndex + 1) % providers.size
+    val provider = providers.get(newIndex)
+    lastUsedMap.set(name, provider.id)
+    return provider
+  }
+
+  fun getAll(organizationId: Long): List<LlmProviderDto> {
+    return llmProviderRepository.getAll(organizationId).map { it.toDto() }
+  }
+
+  fun <T> repeatWhileProvidersRateLimited(
+    organizationId: Long,
+    provider: String,
+    priority: LlmProviderPriority?,
+    callback: (provider: LlmProviderDto) -> T,
+  ): T {
+    var lastError: Exception? = null
+
+    // attempt 3 times to find non-rate-limited provider
+    for (i in 0..3) {
```

Comment:
there is also repeat method: https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/repeat.html